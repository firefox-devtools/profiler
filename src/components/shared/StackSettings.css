/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

.stackSettings {
  display: flex;
  overflow: auto hidden;
  height: 25px;
  flex-flow: row nowrap;
  align-items: stretch;
  padding: 0;
  scrollbar-width: none;
  white-space: nowrap;
}

.stackSettingsSelect {
  /* Create a new stacking context for the select using a position:relative. This way
   * the focus ring around the select breaks out of its surrounding container. */
  position: relative;
  height: 24px;
  color: inherit; /* important so that the text is still legible when the OS is configured in dark mode. */
  font-size: inherit;
}
