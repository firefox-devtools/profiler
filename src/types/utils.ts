/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

/**
 * This type serves as documentation for how an array is meant to be used, but does
 * not support type checking. We often use an Array instead of a Map to translate
 * one type of index into another type of index. This is similar to how we use the
 * Map<K,V> type, but with the Array.
 */
export type IndexedArray<_IndexType, Value> = Array<Value>;

export type MixedObject = { [key: string]: unknown };
