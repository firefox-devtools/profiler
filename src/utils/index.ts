/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import type {
  Marker,
  Milliseconds,
  StartEndRange,
  KeyboardModifiers,
} from 'firefox-profiler/types';
import type {
  MouseEvent as SyntheticMouseEvent,
  KeyboardEvent as SyntheticKeyboardEvent,
} from 'react';

/**
 * Firefox has issues switching quickly between fill style colors, as the CSS color
 * is fully parsed each time it is set. As a mitigation, provide a class that only
 * switches the color when it's really needed.
 */
export class FastFillStyle {
  _ctx: CanvasRenderingContext2D;
  _previousFillColor: string;

  constructor(ctx: CanvasRenderingContext2D) {
    this._ctx = ctx;
    this._previousFillColor = '';
  }

  set(fillStyle: string) {
    if (fillStyle !== this._previousFillColor) {
      this._ctx.fillStyle = fillStyle;
      this._previousFillColor = fillStyle;
    }
  }
}

/**
 * Perform a simple shallow object equality check.
 */
export function objectShallowEquals<
  A extends Record<string, any>,
  B extends Record<string, any>,
>(a: A, b: B): boolean {
  let aLength = 0;
  let bLength = 0;
  for (const key in a) {
    if (Object.prototype.hasOwnProperty.call(a, key)) {
      aLength++;
      if ((a as any)[key] !== (b as any)[key]) {
        return false;
      }
    }
  }
  for (const key in b) {
    if (Object.prototype.hasOwnProperty.call(b, key)) {
      bLength++;
    }
  }
  return aLength === bLength;
}

/**
 * Don't completely trust URLs that get displayed to the screen. These could
 * be extraordinarily long data-uris. This function could also be updated to
 * return a nicely formatted React span.
 */
export function displayNiceUrl(rawUrl: string): string {
  if (rawUrl.length < 200) {
    return rawUrl;
  }
  return `${rawUrl.slice(0, 150)}...${rawUrl.slice(-20)}`;
}

/**
 * This function handles the logic of converting both Interval and Instant markers
 * into a range selection. For instant markers, it creates a range that is 5%
 * of the range selection (typically the committed range).
 */
export function getStartEndRangeForMarker(
  rangeStart: Milliseconds,
  rangeEnd: Milliseconds,
  marker: Marker
): StartEndRange {
  let start = marker.start;
  let end = marker.end;

  if (end === null) {
    const delta = (rangeEnd - rangeStart) / 40;
    start = marker.start - delta;
    end = marker.start + delta;
  }

  return { start, end };
}

/**
 * This logic is shared between multiple components, but it is used to determine how a
 * track gets selected, based on the modifiers used.
 */
export function getTrackSelectionModifiers(
  event:
    | MouseEvent
    | KeyboardEvent
    | SyntheticMouseEvent<any>
    | SyntheticKeyboardEvent<any>
): KeyboardModifiers {
  return {
    ctrlOrMeta: (event.ctrlKey || event.metaKey) && !event.altKey,
    shift: event.shiftKey && !event.altKey,
  };
}

export function countPositiveValues(arr: Array<number>): number {
  let count = 0;
  for (let i = 0; i < arr.length; i++) {
    if (arr[i] > 0) {
      count++;
    }
  }
  return count;
}

/**
 * Find the highest number among the map values and return its key.
 * Returns undefined if the map is empty.
 * If multiple entries with the highest value exist, it returns the key of the
 * first encountered highest value.
 */
export function mapGetKeyWithMaxValue<K>(map: Map<K, number>): K | undefined {
  let maxValue = -Infinity;
  let keyForMaxValue;
  for (const [key, value] of map) {
    if (value > maxValue) {
      maxValue = value;
      keyForMaxValue = key;
    }
  }
  return keyForMaxValue;
}
