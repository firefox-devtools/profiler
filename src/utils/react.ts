/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// This file contains some functions that might be interesting when debugging
// react code.

// This function keeps track of the props of a React component and logs the
// changes between two calls. This is best used in render() or
// componentDidUpdate() like this:
// import { logPropChanges } from 'firefox-profiler/utils/react';
// ...
// render() {
//   logPropChanges(this);
// }
export function logPropChanges(component: any) {
  if (!component.__DEBUG__oldProps) {
    component.__DEBUG__oldProps = component.props;
    return;
  }

  for (const [prop, value] of Object.entries(component.props)) {
    const oldValue = component.__DEBUG__oldProps[prop];
    if (oldValue !== value) {
      console.log(
        `prop "${prop}" changed: old value is`,
        oldValue,
        ', new value is',
        value
      );
    }
  }

  component.__DEBUG__oldProps = component.props;
}