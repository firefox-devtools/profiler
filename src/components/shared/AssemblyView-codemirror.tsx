/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
/**
 * This module wraps all the interaction with the CodeMirror API into a
 * AssemblyViewEditor class.
 *
 * This module is intended to be imported asynchronously, so that all the
 * CodeMirror code can be split into a separate bundle chunk.
 *
 * This file implements the following features:
 *  - Display assembly code.
 *  - Display a gutter with:
 *    - "Total" timings for each instruction
 *    - "Self" timings for each instruction
 *    - The address for each instruction
 *  - Highlight assembly code lines which have a non-zero timing, by applying
 *    a cm-nonZeroLine class to them. This highlight line goes across the entire
 *    width of the editor, it covers both the gutter and the main area.
 */
import { EditorView, gutter } from '@codemirror/view';
import { EditorState, StateField, StateEffect } from '@codemirror/state';
import { syntaxHighlighting } from '@codemirror/language';
import { classHighlighter } from '@lezer/highlight';
import clamp from 'clamp';

import {
  AddressTimings,
  Address,
  LineTimings,
  LineNumber,
  NativeSymbolInfo,
  DecodedInstruction,
} from 'firefox-profiler/types';

import { bisectionRight } from 'firefox-profiler/utils/bisect';
import {
  timingsExtension,
  updateTimingsEffect,
  StringMarker,
} from 'firefox-profiler/utils/codemirror-shared';

// An "effect" is like a redux action. This effect is used to replace the value
// of the state field addressToLineMapField.
const updateAddressToLineMapEffect = StateEffect.define<AddressToLineMap>();

// This "state field" stores the current AddressToLineMap. This field allows the
// instructionAddressGutter to map line numbers to addresses.
const addressToLineMapField = StateField.define<AddressToLineMap>({
  create() {
    return new AddressToLineMap([]);
  },
  update(instructionAddresses, transaction) {
    // Get the new value from an effect in the transaction.
    let newSortedAddresses = instructionAddresses;
    for (const effect of transaction.effects) {
      if (effect.is(updateAddressToLineMapEffect)) {
        newSortedAddresses = effect.value;
      }
    }
    return newSortedAddresses;
  },
});

// A gutter which displays the address of each instruction.
const instructionAddressGutter = gutter({
  class: 'cm-instruction-address-gutter',

  // Returns a gutter marker for this line, or null.
  lineMarker(view, line) {
    const lineNumber = view.state.doc.lineAt(line.from).number;
    const map = view.state.field(addressToLineMapField);
    const address = map.lineToAddress(lineNumber);
    return address !== null
      ? new StringMarker(`0x${address.toString(16)}`)
      : null;
  },

  // Returns true if the update affects the instruction addresses in the gutter.
  lineMarkerChange(update) {
    return update.transactions.some((t) =>
      t.effects.some((e) => e.is(updateAddressToLineMapEffect))
    );
  },
});

function instructionsToText(assemblyCode: DecodedInstruction[]): string {
  return assemblyCode.map((instr) => instr.decodedString).join('\n');
}

/**
 * This map is used to convert between instruction addresses and editor line
 * numbers.
 */
class AddressToLineMap {
  // The address of each instruction. This stays constant for the entire lifetime
  // of this AddressToLineMap instance.
  //
  // _instructionAddresses[0] contains the address of the instruction which is
  // displayed in line 1. (Line numbers are 1-based.)
  //
  // The addresses need to be ordered from low to high, so that the binary search
  // works.
  _instructionAddresses: Address[];

  constructor(instructionAddresses: Address[]) {
    this._instructionAddresses = instructionAddresses;
  }

  // Find the line which displays the instruction which covers `address`.
  // `address` doesn't need to be a perfect match for the instruction address;
  // for example, in the example below, address 0x10e4 is mapped to line 3:
  //
  // 1: 0x10da: mov r14, rdi
  // 2: 0x10dd: mov rdi, rsi
  // 3: 0x10e0: call _malloc_usable_size
  // 4: 0x10e5: test rax, rax
  // 5: 0x10e8: je loc_10f6
  addressToLine(address: Address): LineNumber | null {
    const insertionIndex = bisectionRight(this._instructionAddresses, address);
    if (insertionIndex === 0) {
      // address < instructionAddresses[0]
      return null;
    }

    const elementIndex = insertionIndex - 1;
    const lineNumber = elementIndex + 1;
    return lineNumber;
  }

  // Return the address of the instruction which is displayed in line `lineNumber`.
  lineToAddress(lineNumber: LineNumber): Address | null {
    if (lineNumber < 1 || lineNumber > this._instructionAddresses.length) {
      return null;
    }

    const elementIndex = lineNumber - 1;
    return this._instructionAddresses[elementIndex];
  }
}

function getInstructionAddresses(
  assemblyCode: DecodedInstruction[]
): Address[] {
  return assemblyCode.map((instr) => instr.address);
}

// Convert AddressTimings to LineTimings with the help of an AddressToLineMap.
function addressTimingsToLineTimings(
  addressTimings: AddressTimings,
  map: AddressToLineMap
): LineTimings {
  const totalLineHits = new Map();
  for (const [address, hitCount] of addressTimings.totalAddressHits) {
    const line = map.addressToLine(address);
    if (line !== null) {
      const currentHitCount = totalLineHits.get(line) ?? 0;
      totalLineHits.set(line, currentHitCount + hitCount);
    }
  }

  const selfLineHits = new Map();
  for (const [address, hitCount] of addressTimings.selfAddressHits) {
    const line = map.addressToLine(address);
    if (line !== null) {
      const currentHitCount = selfLineHits.get(line) ?? 0;
      selfLineHits.set(line, currentHitCount + hitCount);
    }
  }

  return { totalLineHits, selfLineHits };
}

export class AssemblyViewEditor {
  _view: EditorView;
  _addressToLineMap: AddressToLineMap;
  _addressTimings: AddressTimings;

  // Create a CodeMirror editor and add it as a child element of domParent.
  constructor(
    initialAssemblyCode: DecodedInstruction[],
    nativeSymbol: NativeSymbolInfo,
    addressTimings: AddressTimings,
    domParent: Element
  ) {
    this._addressToLineMap = new AddressToLineMap(
      getInstructionAddresses(initialAssemblyCode)
    );
    this._addressTimings = addressTimings;
    let state = EditorState.create({
      doc: instructionsToText(initialAssemblyCode),
      extensions: [
        timingsExtension,
        addressToLineMapField,
        instructionAddressGutter,
        syntaxHighlighting(classHighlighter),
        EditorState.readOnly.of(true),
        EditorView.editable.of(false),
      ],
    });
    const lineTimings = addressTimingsToLineTimings(
      this._addressTimings,
      this._addressToLineMap
    );
    state = state.update({
      effects: [
        updateAddressToLineMapEffect.of(this._addressToLineMap),
        updateTimingsEffect.of(lineTimings),
      ],
    }).state;
    this._view = new EditorView({
      state,
      parent: domParent,
    });
  }

  setContents(assemblyCode: DecodedInstruction[]) {
    this._addressToLineMap = new AddressToLineMap(
      getInstructionAddresses(assemblyCode)
    );
    const lineTimings = addressTimingsToLineTimings(
      this._addressTimings,
      this._addressToLineMap
    );
    // The CodeMirror way of replacing the entire contents is to insert new text
    // and overwrite the full range of existing text.
    const text = instructionsToText(assemblyCode);
    this._view.dispatch(
      this._view.state.update({
        changes: {
          insert: text,
          from: 0,
          to: this._view.state.doc.length,
        },
      })
    );
    this._view.dispatch({
      effects: [
        updateAddressToLineMapEffect.of(this._addressToLineMap),
        updateTimingsEffect.of(lineTimings),
      ],
    });
  }

  setTimings(addressTimings: AddressTimings) {
    // Update the value of the timings field by dispatching an updateTimingsEffect.
    this._addressTimings = addressTimings;
    const lineTimings = addressTimingsToLineTimings(
      this._addressTimings,
      this._addressToLineMap
    );
    this._view.dispatch({
      effects: updateTimingsEffect.of(lineTimings),
    });
  }

  scrollToLine(lineNumber: number) {
    // Clamp the line number to the document's line count.
    lineNumber = clamp(lineNumber, 1, this._view.state.doc.lines);

    // Convert the line number into a position.
    const pos = this._view.state.doc.line(lineNumber).from;
    // Dispatch the scroll action.
    this._view.dispatch({
      effects: EditorView.scrollIntoView(pos, { y: 'start', yMargin: 0 }),
    });
    // Trigger a measure flush, to work around
    // https://github.com/codemirror/codemirror.next/issues/676
    this._view.coordsAtPos(0);
  }

  scrollToAddress(address: Address) {
    const lineNumber = this._addressToLineMap.addressToLine(address);
    if (lineNumber !== null) {
      this.scrollToLine(lineNumber);
    }
  }

  scrollToAddressWithSpaceOnTop(address: Address, topSpaceLines: number) {
    const lineNumber = this._addressToLineMap.addressToLine(address);
    if (lineNumber !== null) {
      this.scrollToLine(lineNumber - topSpaceLines);
    }
  }
}
