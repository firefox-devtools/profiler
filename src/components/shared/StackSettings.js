/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import React, { PureComponent } from 'react';
import { Localized } from '@fluent/react';

import {
  changeInvertCallstack,
  changeCallTreeSearchString,
  changeShowUserTimings,
  changeStackChartSameWidths,
} from 'firefox-profiler/actions/profile-view';
import {
  getInvertCallstack,
  getSelectedTab,
  getShowUserTimings,
  getStackChartSameWidths,
  getCurrentSearchString,
} from 'firefox-profiler/selectors/url-state';
import { getProfileUsesMultipleStackTypes } from 'firefox-profiler/selectors/profile';
import { PanelSearch } from './PanelSearch';
import { StackImplementationSetting } from './StackImplementationSetting';
import { CallTreeStrategySetting } from './CallTreeStrategySetting';

import explicitConnect, {
  type ConnectedProps,
} from 'firefox-profiler/utils/connect';
import { selectedThreadSelectors } from 'firefox-profiler/selectors/per-thread';

import './PanelSettingsList.css';
import './StackSettings.css';

type OwnProps = {|
  +hideInvertCallstack?: true,
|};

type StateProps = {|
  +selectedTab: string,
  +allowSwitchingStackType: boolean,
  +invertCallstack: boolean,
  +showUserTimings: boolean,
  +stackChartSameWidths: boolean,
  +currentSearchString: string,
  +hasUsefulJsAllocations: boolean,
  +hasUsefulNativeAllocations: boolean,
|};

type DispatchProps = {|
  +changeInvertCallstack: typeof changeInvertCallstack,
  +changeShowUserTimings: typeof changeShowUserTimings,
  +changeCallTreeSearchString: typeof changeCallTreeSearchString,
  +changeStackChartSameWidths: typeof changeStackChartSameWidths,
|};

type Props = ConnectedProps<OwnProps, StateProps, DispatchProps>;

class StackSettingsImpl extends PureComponent<Props> {
  _onInvertCallstackClick = (e: SyntheticEvent<HTMLInputElement>) => {
    this.props.changeInvertCallstack(e.currentTarget.checked);
  };

  _onShowUserTimingsClick = (e: SyntheticEvent<HTMLInputElement>) => {
    this.props.changeShowUserTimings(e.currentTarget.checked);
  };

  _onUseStackChartSameWidths = (e: SyntheticEvent<HTMLInputElement>) => {
    this.props.changeStackChartSameWidths(e.currentTarget.checked);
  };

  _onSearch = (value: string) => {
    this.props.changeCallTreeSearchString(value);
  };

  render() {
    const {
      allowSwitchingStackType,
      invertCallstack,
      selectedTab,
      showUserTimings,
      stackChartSameWidths,
      hideInvertCallstack,
      currentSearchString,
      hasUsefulJsAllocations,
      hasUsefulNativeAllocations,
    } = this.props;

    const hasAllocations = hasUsefulJsAllocations || hasUsefulNativeAllocations;

    return (
      <div className="stackSettings">
        <ul className="panelSettingsList">
          {allowSwitchingStackType ? (
            <li className="panelSettingsListItem">
              <StackImplementationSetting />
            </li>
          ) : null}
          {hasAllocations ? (
            <li className="panelSettingsListItem">
              <CallTreeStrategySetting />
            </li>
          ) : null}
          {hideInvertCallstack && selectedTab !== 'stack-chart' ? null : (
            <li className="panelSettingsListItem">
              {hideInvertCallstack ? null : (
                <label className="photon-label photon-label-micro photon-label-horiz-padding">
                  <input
                    type="checkbox"
                    className="photon-checkbox photon-checkbox-micro stackSettingsCheckbox"
                    onChange={this._onInvertCallstackClick}
                    checked={invertCallstack}
                  />
                  <Localized
                    id="StackSettings--invert-call-stack"
                    attrs={{ title: true }}
                  >
                    <span>Invert call stack</span>
                  </Localized>
                </label>
              )}
              {selectedTab !== 'stack-chart' ? null : (
                <>
                  <label className="photon-label photon-label-micro photon-label-horiz-padding">
                    <input
                      type="checkbox"
                      className="photon-checkbox photon-checkbox-micro stackSettingsCheckbox"
                      onChange={this._onShowUserTimingsClick}
                      checked={showUserTimings}
                    />
                    <Localized id="StackSettings--show-user-timing">
                      Show user timing
                    </Localized>
                  </label>
                  <label className="photon-label photon-label-micro photon-label-horiz-padding">
                    <input
                      type="checkbox"
                      className="photon-checkbox photon-checkbox-micro stackSettingsCheckbox"
                      onChange={this._onUseStackChartSameWidths}
                      checked={stackChartSameWidths}
                    />
                    <Localized id="StackSettings--use-stack-chart-same-widths">
                      Use the same width for each stack
                    </Localized>
                  </label>
                </>
              )}
            </li>
          )}
        </ul>
        <Localized
          id="StackSettings--panel-search"
          attrs={{ label: true, title: true }}
        >
          <PanelSearch
            className="stackSettingsSearchField"
            label="Filter stacks:"
            title="Only display stacks which contain a function whose name matches this substring"
            currentSearchString={currentSearchString}
            onSearch={this._onSearch}
          />
        </Localized>
      </div>
    );
  }
}

export const StackSettings = explicitConnect<
  OwnProps,
  StateProps,
  DispatchProps,
>({
  mapStateToProps: (state) => ({
    allowSwitchingStackType: getProfileUsesMultipleStackTypes(state),
    invertCallstack: getInvertCallstack(state),
    selectedTab: getSelectedTab(state),
    showUserTimings: getShowUserTimings(state),
    stackChartSameWidths: getStackChartSameWidths(state),
    currentSearchString: getCurrentSearchString(state),
    hasUsefulJsAllocations:
      selectedThreadSelectors.getHasUsefulJsAllocations(state),
    hasUsefulNativeAllocations:
      selectedThreadSelectors.getHasUsefulNativeAllocations(state),
  }),
  mapDispatchToProps: {
    changeInvertCallstack,
    changeCallTreeSearchString,
    changeShowUserTimings,
    changeStackChartSameWidths,
  },
  component: StackSettingsImpl,
});
