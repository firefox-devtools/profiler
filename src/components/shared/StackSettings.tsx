/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import { PureComponent } from 'react';
import { Localized } from '@fluent/react';

import {
  changeInvertCallstack,
  changeCallTreeSearchString,
  changeShowUserTimings,
  changeStackChartSameWidths,
} from 'firefox-profiler/actions/profile-view';
import {
  getInvertCallstack,
  getSelectedTab,
  getShowUserTimings,
  getStackChartSameWidths,
  getCurrentSearchString,
} from 'firefox-profiler/selectors/url-state';
import { getProfileUsesMultipleStackTypes } from 'firefox-profiler/selectors/profile';
import { PanelSearch } from './PanelSearch';
import { StackImplementationSetting } from './StackImplementationSetting';
import { CallTreeStrategySetting } from './CallTreeStrategySetting';

import explicitConnect, {
  type ConnectedProps,
} from 'firefox-profiler/utils/connect';
import { selectedThreadSelectors } from 'firefox-profiler/selectors/per-thread';

import './PanelSettingsList.css';
import './StackSettings.css';

type OwnProps = {
  readonly hideInvertCallstack?: true;
};

type StateProps = {
  readonly selectedTab: string;
  readonly allowSwitchingStackType: boolean;
  readonly invertCallstack: boolean;
  readonly showUserTimings: boolean;
  readonly stackChartSameWidths: boolean;
  readonly currentSearchString: string;
  readonly hasUsefulJsAllocations: boolean;
  readonly hasUsefulNativeAllocations: boolean;
};

type DispatchProps = {
  readonly changeInvertCallstack: typeof changeInvertCallstack;
  readonly changeShowUserTimings: typeof changeShowUserTimings;
  readonly changeCallTreeSearchString: typeof changeCallTreeSearchString;
  readonly changeStackChartSameWidths: typeof changeStackChartSameWidths;
};

type Props = ConnectedProps<OwnProps, StateProps, DispatchProps>;

class StackSettingsImpl extends PureComponent<Props> {
  _onInvertCallstackClick = (e: React.ChangeEvent<HTMLInputElement>) => {
    this.props.changeInvertCallstack(e.currentTarget.checked);
  };

  _onShowUserTimingsClick = (e: React.ChangeEvent<HTMLInputElement>) => {
    this.props.changeShowUserTimings(e.currentTarget.checked);
  };

  _onUseStackChartSameWidths = (e: React.ChangeEvent<HTMLInputElement>) => {
    this.props.changeStackChartSameWidths(e.currentTarget.checked);
  };

  _onSearch = (value: string) => {
    this.props.changeCallTreeSearchString(value);
  };

  override render() {
    const {
      allowSwitchingStackType,
      invertCallstack,
      selectedTab,
      showUserTimings,
      stackChartSameWidths,
      hideInvertCallstack,
      currentSearchString,
      hasUsefulJsAllocations,
      hasUsefulNativeAllocations,
    } = this.props;

    const hasAllocations = hasUsefulJsAllocations || hasUsefulNativeAllocations;

    return (
      <div className="stackSettings">
        <ul className="panelSettingsList">
          {allowSwitchingStackType ? (
            <li className="panelSettingsListItem">
              <StackImplementationSetting />
            </li>
          ) : null}
          {hasAllocations ? (
            <li className="panelSettingsListItem">
              <CallTreeStrategySetting />
            </li>
          ) : null}
          {hideInvertCallstack && selectedTab !== 'stack-chart' ? null : (
            <li className="panelSettingsListItem">
              {hideInvertCallstack ? null : (
                <label className="photon-label photon-label-micro photon-label-horiz-padding">
                  <input
                    type="checkbox"
                    className="photon-checkbox photon-checkbox-micro stackSettingsCheckbox"
                    onChange={this._onInvertCallstackClick}
                    checked={invertCallstack}
                  />
                  <Localized
                    id="StackSettings--invert-call-stack"
                    attrs={{ title: true }}
                  >
                    <span>Invert call stack</span>
                  </Localized>
                </label>
              )}
              {selectedTab !== 'stack-chart' ? null : (
                <>
                  <label className="photon-label photon-label-micro photon-label-horiz-padding">
                    <input
                      type="checkbox"
                      className="photon-checkbox photon-checkbox-micro stackSettingsCheckbox"
                      onChange={this._onShowUserTimingsClick}
                      checked={showUserTimings}
                    />
                    <Localized id="StackSettings--show-user-timing">
                      Show user timing
                    </Localized>
                  </label>
                  <label className="photon-label photon-label-micro photon-label-horiz-padding">
                    <input
                      type="checkbox"
                      className="photon-checkbox photon-checkbox-micro stackSettingsCheckbox"
                      onChange={this._onUseStackChartSameWidths}
                      checked={stackChartSameWidths}
                    />
                    <Localized id="StackSettings--use-stack-chart-same-widths">
                      Use the same width for each stack
                    </Localized>
                  </label>
                </>
              )}
            </li>
          )}
        </ul>
        <Localized
          id="StackSettings--panel-search"
          attrs={{ label: true, title: true }}
        >
          <PanelSearch
            className="stackSettingsSearchField"
            label="Filter stacks:"
            title="Only display stacks which contain a function whose name matches this substring"
            currentSearchString={currentSearchString}
            onSearch={this._onSearch}
          />
        </Localized>
      </div>
    );
  }
}

export const StackSettings = explicitConnect<
  OwnProps,
  StateProps,
  DispatchProps
>({
  mapStateToProps: (state) => ({
    allowSwitchingStackType: getProfileUsesMultipleStackTypes(state),
    invertCallstack: getInvertCallstack(state),
    selectedTab: getSelectedTab(state),
    showUserTimings: getShowUserTimings(state),
    stackChartSameWidths: getStackChartSameWidths(state),
    currentSearchString: getCurrentSearchString(state),
    hasUsefulJsAllocations:
      selectedThreadSelectors.getHasUsefulJsAllocations(state),
    hasUsefulNativeAllocations:
      selectedThreadSelectors.getHasUsefulNativeAllocations(state),
  }),
  mapDispatchToProps: {
    changeInvertCallstack,
    changeCallTreeSearchString,
    changeShowUserTimings,
    changeStackChartSameWidths,
  },
  component: StackSettingsImpl,
});
