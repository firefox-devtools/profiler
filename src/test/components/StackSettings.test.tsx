/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
import { fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';

import { render, act } from 'firefox-profiler/test/fixtures/testing-library';
import { StackSettings } from '../../components/shared/StackSettings';
import { storeWithProfile } from '../fixtures/stores';
import {
  getImplementationFilter,
  getCurrentSearchString,
} from '../../selectors/url-state';
import { fireFullClick } from '../fixtures/utils';

describe('StackSettings', function () {
  function setup() {
    jest.useFakeTimers();
    const store = storeWithProfile();

    const renderResult = render(
      <Provider store={store}>
        <StackSettings />
      </Provider>
    );

    return {
      ...renderResult,
      ...store,
    };
  }

  it('matches the snapshot', () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });

  /**
   * Get around the type constraints of refining an HTMLElement into a radio input.
   */
  function getCheckedState(element: HTMLElement): unknown {
    return (element as any).checked;
  }

  it('can change the implementation filter to script', async function () {
    const { getByLabelText, getState } = setup();
    expect(getImplementationFilter(getState())).toEqual('combined');
    const radioButton = getByLabelText(/Script/);

    fireFullClick(radioButton);

    expect(getCheckedState(radioButton)).toBe(true);
    expect(getImplementationFilter(getState())).toEqual('js');
  });

  it('can change the implementation filter to Native', function () {
    const { getByLabelText, getState } = setup();
    expect(getImplementationFilter(getState())).toEqual('combined');
    const radioButton = getByLabelText(/Native/);

    fireFullClick(radioButton);

    expect(getCheckedState(radioButton)).toBe(true);
    expect(getImplementationFilter(getState())).toEqual('cpp');
  });

  it('can change the implementation filter to All frames', function () {
    const { getByLabelText, getState } = setup();
    fireFullClick(getByLabelText(/Native/));
    expect(getImplementationFilter(getState())).toEqual('cpp');
    const radioButton = getByLabelText(/All frames/);

    fireFullClick(radioButton);

    expect(getCheckedState(radioButton)).toBe(true);
    expect(getImplementationFilter(getState())).toEqual('combined');
  });

  it('can change the search', function () {
    const { getByLabelText, getState } = setup();
    expect(getCurrentSearchString(getState())).toEqual('');
    const searchText = 'some search';
    const input: HTMLInputElement = getByLabelText(/Filter stacks/) as any;

    fireEvent.change(input, {
      target: { value: searchText },
    });

    act(() => jest.runAllTimers());
    expect(getCurrentSearchString(getState())).toEqual(searchText);
    expect(input.value).toEqual(searchText);
  });
});
