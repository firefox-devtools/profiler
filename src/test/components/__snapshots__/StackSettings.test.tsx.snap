// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`StackSettings matches the snapshot 1`] = `
<div>
  <div
    class="stackSettings"
  >
    <ul
      class="panelSettingsList"
    >
      <li
        class="panelSettingsListItem"
      >
        <input
          checked=""
          class="photon-radio photon-radio-micro"
          id="implementation-radio-combined"
          title="Do not filter the stack frames"
          type="radio"
          value="combined"
        />
        <label
          class="photon-label photon-label-micro photon-label-horiz-padding"
          for="implementation-radio-combined"
          title="Do not filter the stack frames"
        >
          All frames
        </label>
        <input
          class="photon-radio photon-radio-micro"
          id="implementation-radio-js"
          title="Show only the stack frames related to script execution"
          type="radio"
          value="js"
        />
        <label
          class="photon-label photon-label-micro photon-label-horiz-padding"
          for="implementation-radio-js"
          title="Show only the stack frames related to script execution"
        >
          Script
        </label>
        <input
          class="photon-radio photon-radio-micro"
          id="implementation-radio-cpp"
          title="Show only the stack frames for native code"
          type="radio"
          value="cpp"
        />
        <label
          class="photon-label photon-label-micro photon-label-horiz-padding"
          for="implementation-radio-cpp"
          title="Show only the stack frames for native code"
        >
          Native
        </label>
      </li>
      <li
        class="panelSettingsListItem"
      >
        <label
          class="photon-label photon-label-micro photon-label-horiz-padding"
        >
          <input
            class="photon-checkbox photon-checkbox-micro stackSettingsCheckbox"
            type="checkbox"
          />
          <span
            title="Sort by the time spent in a call node, ignoring its children."
          >
            Invert call stack
          </span>
        </label>
      </li>
    </ul>
    <div
      class="panelSearchField stackSettingsSearchField"
    >
      <label
        class="panelSearchFieldLabel"
      >
        Filter stacks: 
        <form
          class="idleSearchField panelSearchFieldInput"
        >
          <input
            class="idleSearchFieldInput photon-input"
            name="search"
            placeholder="Enter filter terms"
            required=""
            title="Only display stacks which contain a function whose name matches this substring"
            type="search"
            value=""
          />
          <input
            class="idleSearchFieldButton"
            tabindex="-1"
            type="reset"
          />
        </form>
      </label>
      <div
        class="panelSearchFieldIntroduction isHidden"
      >
        Did you know you can use the comma (,) to search using several terms?
      </div>
    </div>
  </div>
</div>
`;
